### ISSUE TITLE
Allow to extend configuration objects instead of cascading environments config

### ISSUE THREAD

**Issue by mzgol** [eslint/eslint#1637]
I know extending configurations similarly to what jsHint does was argued early in the project and the current solution is to just propagate options down the tree. This has some unwanted side effects, though.

A lot of JavaScript projects have structure loosely resembling this one:
```
├─┬ app
│ └── files for the browser
├─┬ dist
│ └── files for the browser
├─┬ test
│ ├── unit tests - executed in a browser
│ └── e2e tests - executed in Node
├── Gruntfile.js - for Node
└── karma.conf.js - for Node
```
This means that there has to be the `.eslintrc` file contained in the top project scope and it has to have the `node` environment enabled. But that means `"node": true` is also the default for `app`/`dist`/`test/unit`/etc. and has to be explicitly disabled in each of them separately which is easy to forget about (or even not think about it being needed).

OTH, if instead of cascading environments one could extend a file with base config for the whole project, all these workarounds would be unnecessary.

**Comment by nzakas** [comment]
I'm just not sure extending configs is useful enough to add that complexity on top of what we already have. The cascading configuration turned out to be very complex and took a while to get correct but overall feedback has been positive. Inserting another way of modifying the config into what's already a complex system is something I'd like to avoid unless it's really necessary.

**Comment by mzgol** [comment]
I guess my main problem is that if I have Node code at the root of the project and browser code in subdirectories, then this:
```json
{
    "env": {
        "browser": true
    }
}
```
is not enough and I have to do this:
```json
{
    "env": {
        "browser": true,
        "node": false
    }
}
```

I've just discovered it when working on a large-ish project at my company; it scared me a little that all the browser code was linted assuming Node settings in addition to browser ones.

**Comment by nzakas** [comment]
I'll leave this open to see if anyone wants to investigate. Right now we're in the midst of trying to get ES6 and JSX support out, so I won't have time to really look at this until that work is done first.

**Comment by mzgol** [comment]
Sure, I understand.

**Comment by gcochard** [comment]
I have the same type of layout. It's definitely annoying having to override the `"node": true` everywhere.

What would your ideal config files look like, and where would they be stored? I'd like to get an idea of how to do this properly.

**Comment by gcochard** [comment]
@mzgol My main concern is how to extend the base while in the root directory of the project.

**Comment by mzgol** [comment]
@gcochard 
> @mzgol My main concern is how to extend the base while in the root directory of the project.

With jsHint I used the pattern of naming the base file `.jshintrc-base` (i.e. sth not automatically picked up by jsHint) and then extending it in all `.jshintrc`s.

I like propagating linting rules, the problem are propagating environments.

For my purposes it'd be enough if by default setting `browser` to `true` was setting `node` to false and vice-versa. If one wanted to have both set to `true`, it could be done explicitly. I know it would mean special-casing specific environments which doesn't seem nice but what exactly is a use case of being able to set both `browser` and `node` to `true`? If you write code for both environments, you most likely need the opposite - to have your code assume the presence of neither of them, i.e. both envs set to `false`.

**Comment by gcochard** [comment]
I think it does make sense to have a few mutually-exclusive environments.

I think setting both `browser` and `node` to true would be necessary in a few contrived environments such as when when you write a function in a node source file and pass that to the browser with `function.toString()`. At that point you're probably better off writing the function somewhere else (or not using `toString()` at all...and this might be a strict mode violation anyway, depending on how it's written). This is certainly not a best practice, and I haven't seen it used in practice, but it's possible to do.

Can anyone else come up with a scenario where the browser and node environments might overlap?

**Comment by nzakas** [comment]
This discussion is going a bit too far off topic. We aren't going to change how environments work. I have seen projects that choose to turn on both browser and node environments because the same file is used slightly differently in each environment.

The topic at hand is how to extend configs, so let's focus in on that.

**Comment by mightyiam** [comment]
@nzakas thank you for pointing me here.

I'll be very happy to share my requirement/use case for this and will also be happy to hear your thoughts about it, so here it goes.

My requirement for this comes from the project named [standard](https://github.com/feross/standard), where we aim to allow willing project maintainers to use a widely–used JavaScript code style standard.

Currently, the tool works by providing a bin calles `standard`, which runs both ESLint and JSCS in parallel, with the standard's rules provided to these tools in configuration files.

This allows usage in test scripts like in `npm test` and in on–change watch scripts.

But it doesn't allow use through ESLint and JSCS, when they are used by themselves, or through another tool (like an editor).

Actually, we can always call, for example, ESLint, with the command–line argument that direct it to our `node_modules/standard/rc/.eslintrc`, and that will work.

But, [we would like it](feross/standard/issues/33) to work by default—as in without any arguments—and through tools (like editors), as well.

This is why I thought that an include rule (a possible means of addressing this here issue #1637), will help.

My thought is that when our package, `standard`, is installed, it will, via an npm `install`/`postinstall` script, add an ESLint rule to `package.json`, that is actually an include rule, that directs ESLint to our `node_modules/standard/rc/.eslintrc`.

I will be very happy if, whichever way you choose to implement #1637, it will take this use case into account.

Thank you!

**Comment by nzakas** [comment]
If you want to implement `extends`, I'd say go for it.

**Comment by thefotios** [comment]
I've been thinking about this as well and have a similar issue. One idea I had was possible including something in the various `.eslintrc` files like `root = true` which could tell the parser to stop traversing up directories for other files to include. This would allow directory specific files to only be applied against the global/default configs and not take into account anything else in your directories.

This is a similar approach to how `editorconfig` works.

**Comment by nzakas** [comment]
@thefotios that's out of scope for this issue. You can accomplish the same thing once this is fine by only having one eslintrc file in your project.

**Comment by mightyiam** [comment]
Thank you very much, @rexxars.

**Comment by G3z** [comment]
i have a problem with extension and i think it's related to this thread.

with 0.18 i had an `.eslintrc` in my project with globals 
```json
{
	"globals": {
		"_": true,
		"baseUrl": true,
		"cordova": true,
		"db": true,
		"router": true,
		"server": true
	}
}
```

and a global `.eslintrc` in my home directory
```json
{
	"parser": "babel-eslint",
	"plugins": [
		"react"
	],
	"env": {
		"browser": true,
		"node": true
	},
	"rules": {
		"quotes": [0],
		"eol-last": [0],
		"no-mixed-requires": [0],
		"no-underscore-dangle": [0],
		"eqeqeq": [1] ,
		"curly": [1],
		"semi": [1],
		"strict": [0],
		"no-undef": [1],
		"no-multiple-empty-lines": [1]
	}
}
```

and it worked.
now the local one obfuscate the global
i even tried `"extends": "/Users/g3z/.eslintrc"` but it doesn't work: any idea ?

**Comment by nzakas** [comment]
This feature hasn't been released yet.

**Comment by nzakas** [comment]
> now the local one obfuscate the global

I think you mean the local config overrides the global. This is how configuration works in ESLint. See: http://eslint.org/docs/user-guide/configuring

**Comment by G3z** [comment]
sorry i'm pretty sure that this was working before upgrading.
but maybe it's just a coincidence

i have eslint 0.20 on node 0.12 and osx 10.10

i have those 2 files in
`~/Documents/dev/mobile/projectA/.eslintrc` (local)
and the other in
`~/.eslintrc` (global)

i noticed that if global is `~/Documents/.eslintrc`or even `/.eslintrc` it works as expected (global + local)
if global path is exactly `~/.eslintrc` it's ignored.

**Comment by nzakas** [comment]
Sorry, you're not being very clear about the problem. I think now you're saying that `~/.eslintrc` isn't being loaded when you have a local `.eslintrc`? That was a bug fix in 0.20: https://github.com/eslint/eslint/issues/2279

If you have further questions, please open a new issue or contact the mailing list.
